### PR TITLE
[ONNX] Refactor _run_symbolic_function

### PR DESCRIPTION
**Pull request by BowenBao**
* Allows implementing symbolic functions for domains other than `aten`, for example `prim`, in symbolic_opset#.py.
* Allows symbolic function to access extra context if needed, through `SymbolicFunctionState`.
  * Particularly, the `prim::PythonOp` special case can access node without the need of passing node through inputs. Updates will be made downstreams, and in a follow-up PR we will remove the previous workaround in exporter.
* `prim::Loop`, `prim::If`, etc are now moved outside of `_run_symbolic_function` from utils.py, and to symbolic_opset9.py.

Motivation for this change:
- Better maintainability and reducing complexity. Easier to add symbolic for operators, both simple and complex ones (that need additional context), without the former needing to know the existence of the latter.
- The design idea was long outdated. prim ops are no longer rare special cases, and they shouldn't all be handled inside `_run_symbolic_function`. As a result this function becomes too clumsy. There were also prim ops symbolic added in symbolic_opset#.py with signature `prim_[opname]`, creating separation and confusion.

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
> * Allows implementing symbolic functions for domains other than `aten`, for example `prim`, in symbolic_opset#.py.
> * Allows symbolic function to access extra context if needed, through `SymbolicFunctionState`.
>   
>   * Particularly, the `prim::PythonOp` special case can access node without the need of passing node through inputs. Updates will be made downstreams, and in a follow-up PR we will remove the previous workaround in exporter.
> * `prim::Loop`, `prim::If`, etc are now moved outside of `_run_symbolic_function` from utils.py, and to symbolic_opset9.py.

As a curiosity what is the motivation for such change?

**Comment by BowenBao** [comment]
> As a curiosity what is the motivation for such change?

Better maintainability and reducing complexity. Easier to add symbolic for operators, both simple and complex ones (that need additional context), without the former needing to know the existence of the latter.

The design idea was long outdated. It was no longer true that prim ops are rare special cases and they shouldn't all be handled inside `_run_symbolic_function`. As a result this function becomes too clumsy. There were also prim ops symbolic added in symbolic_opset#.py with signature `prim_[opname]`, creating separation and confusion.

**Review by thiagocrepaldi — COMMENTED** [review]
Any reason we don't unify this class approach for aten ops, too? This way we would have a single way implement symbolics: through `class <domain>` classes

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -867,6 +881,39 @@ def _set_training_mode(training_mode):
     global _training_mode
     _training_mode = training_mode
 
+# Symbolic function state.
+class SymbolicFunctionState:
```
[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_helper.py:885]: `class Foo` is the same as `class Foo(object)` in python 3.

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -42,6 +44,18 @@
...
 # concrete shapes. Doing so will make the export symbolics
 # more robust to different graphs.
+#
+# ----------------------------------------------------------------------------------
+# Extra context for Symbolic functions
+# ----------------------------------------------------------------------------------
+#
+# In general, symbolic functions only requires knowing inputs and attributes to
+# the original node. In rare circumstances, extra context may be required.
+# For example, symbolic function for `prim::Loop` needs access to the subblock of
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:54]: For now that `state` is changing though the process of export, the values it holds is different for each symbolic function call. Hence the name. `operator_export_type` is static and global in that regard, so I'm not sure if this is the right place to put it.

**Comment by BowenBao** [comment]
> Any reason we don't unify this class approach for aten ops, too? This way we would have a single way implement symbolics: through `class <domain>` classes

I felt the convention of aten by default for symbolics is commonly accepted now, so being careful in bringing big changes. I agree unifying aten too would be great, it also resolves a lot of builtin shadowing issues, like symbolic for min, max, sum, etc.

File: torch/onnx/symbolic_helper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_helper.py:885]: Yes, semantically the same. I just felt being explicit was a good practice because there is plenty of people still using python2 out there

File: torch/onnx/symbolic_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_helper.py:54]: makes sense

File: torch/onnx/symbolic_helper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_helper.py:885]: I disagree. This code does not support Python 2 and making it seem like it does just adds verbosity and potential for confusion.

+1 for type hints though.

[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:885]: added. fyi PyTorch officially dropped support for python2.

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -867,6 +881,40 @@ def _set_training_mode(training_mode):
     global _training_mode
     _training_mode = training_mode
 
+# Symbolic function state.
+class SymbolicFunctionState(object):
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:885]: I'm sorry to contradict the other reviewer, but please don't write `(object)`. See earlier comment for rationale.

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -867,6 +881,40 @@ def _set_training_mode(training_mode):
...
+    # Skip check due to cannot import IValue from torch._C
+    params_dict = None  # type: ignore[var-annotated]
+    env = None  # type: Optional[Dict[torch._C.Value, torch._C.Value]]
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:888]: please use python 3 style type hints, not comments

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -867,6 +881,40 @@ def _set_training_mode(training_mode):
     global _training_mode
     _training_mode = training_mode
 
+# Symbolic function state.
+class SymbolicFunctionState(object):
+    # Skip check due to cannot import IValue from torch._C
+    params_dict = None  # type: ignore[var-annotated]
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:887]: Are these class variables rather than instance variables on purpose?
It's a bit weird to instantiate this object and then access its class variables.

**Review by garymm — CHANGES_REQUESTED** [review]

File: torch/onnx/symbolic_helper.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:888]: curious what's the rationale? I added python 3 style first but noticed that it was not supported for python < 3.6.

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py]: adjusted

File: torch/onnx/symbolic_helper.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py]: changed to instance variables

File: torch/onnx/symbolic_helper.py (hunk 4 above)
Comment (garymm) [reply torch/onnx/symbolic_helper.py]: It's a more modern, more compact way to write the same thing.
I think PyTorch requires Python >= 3.6 (or at least CI only tests back to 3.6), so there's no reason to use the old comment style.

Comment (thiagocrepaldi) [reply torch/onnx/symbolic_helper.py]: linters and static code analyzer do use type hints. Maybe they also do it for comment style, but I am not sure

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -3322,3 +3258,241 @@ def broadcast_tensors(g, self):
...
+            #     block1():
+            #     -> (%input.1, %embedding_matrix.1)
+            # %26 : int[] = aten::size(%input.53)
+            #
+            # The converted ONNX graph
+            # %10 : Bool(device=cpu) = onnx::Constant[value={0}]()
+            # %14 : Bool(device=cpu) = onnx::Equal(%13, %8)
+            # %15 : Bool(requires_grad=0, device=cpu) = onnx::Constant[value={0}]()
+            # %16 : Long(1, strides=[1], device=cpu) = onnx::Shape(%input.1)
+            input_flag = inputs[0].node()['value'].tolist()
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:3432]:
```suggestion
            input_flag = inputs[0].node()["value"].tolist()
```

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -3322,3 +3258,241 @@ def broadcast_tensors(g, self):
...
+            const_value = all(input_flag) if isinstance(input_flag, list) else bool(input_flag)
+            block_idx = 0 if const_value else 1
+            current_b = list(n.blocks())[block_idx]
+            is_sub_block = True
+            env = torch._C._jit_pass_onnx_block(current_b, block, operator_export_type, env,  # type:ignore[arg-type]
+                                                is_sub_block)
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:3438]: I guess we can just replace `is_sub_block` with `True` here?

File: torch/onnx/symbolic_opset9.py [hunk 8]
```diff
@@ -3322,3 +3258,241 @@ def broadcast_tensors(g, self):
...
+                    raise RuntimeError("The sub block ATen output " + current_b_list[idx] + " is not in env.")
+                onnx_b = env[current_b_list[idx]]
+                final_b_list.append(onnx_b)
+            return final_b_list
+        else:
+            new_op_outputs = g.op("If", *inputs, outputs=n.outputsSize())
+            new_node = new_op_outputs[0].node() if n.outputsSize() > 1 else new_op_outputs.node()
+            for b in n.blocks():
+                new_block = new_node.addBlock()
+                torch._C._jit_pass_onnx_block(b, new_block, operator_export_type, env, is_sub_block)  # type:ignore[arg-type]
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:3454]: and replace `is_sub_block` with `False` here?

File: torch/onnx/symbolic_helper.py [hunk 9]
```diff
@@ -867,6 +881,41 @@ def _set_training_mode(training_mode):
...
+def _clear_symbolic_function_state():
+    global _symbolic_function_state
+    _symbolic_function_state.params_dict = None
+    _symbolic_function_state.env = None
+    _symbolic_function_state.cur_node = None
+    _symbolic_function_state.onnx_block = None
+
+def _get_symbolic_function_state():
+    global _symbolic_function_state
+    assert _symbolic_function_state.params_dict is not None
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:912]: It'd be nice to have a helpful error message. Maybe:

```python
state = _symbolic_function_state
for attr in (state.params_dict, state.env, state.cur_node, state.onnx_block):
  if attr is None:
    raise RuntimeError("symbolic_function_state() called but global _symbolic_function_state is not set. This function should only be called from within a symbolic function.")
return state
```

File: torch/onnx/symbolic_helper.py [hunk 10]
```diff
@@ -867,6 +881,41 @@ def _set_training_mode(training_mode):
     global _training_mode
     _training_mode = training_mode
 
+# Symbolic function state.
+class SymbolicFunctionState:
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_helper.py:886]: And another thought on the name:
Many comments refer to `context` but this class is called `state`. Let's be consistent and pick one word. This makes it clear that there is only a single thing being referred to by both this class and those comments.
Either name this class `_SymbolicFunctionContext` or change the comments. E.g. `Symbolic functions that need extra context` -> `Symbolic functions that need extra state`

File: torch/onnx/symbolic_registry.py [hunk 11]
```diff
@@ -27,21 +27,9 @@ def register_version(domain, version):
...
-            op = ("any", op[1])
-        if op[0] == "_all":
-            op = ("all", op[1])
-        domain_register = domain
-        if op[0].startswith("prim_"):
-            op = (op[0][5:], op[1])
-            domain_register = "prim"
-        if isfunction(op[1]) and not is_registered_op(op[0], domain_register, version):
-            register_op(op[0], op[1], domain_register, version)
+    for domain, op_name, op_func in version_ops:
```
Comment (garymm) [inline torch/onnx/symbolic_registry.py:30]:
```suggestion
    for domain, op_name, op_func in get_ops_in_version(iter_version):
```

File: torch/onnx/symbolic_registry.py [hunk 12]
```diff
@@ -120,11 +125,19 @@ def get_registered_op(opname, domain, version):
         warnings.warn("ONNX export failed. The ONNX domain and/or version are None.")
     global _registry
     if not is_registered_op(opname, domain, version):
+        raise_unsupported_operator_error(domain, opname, version)
+    return _registry[(domain, version)][opname]
+
+def raise_unsupported_operator_error(domain, opname, version):
```
Comment (garymm) [inline torch/onnx/symbolic_registry.py:131]: I suggest we replace this with a class:

```python
class UnsupportedOperatorError(RuntimeError):
```

It can construct the message in its `__init__`.

That makes the code at the call site much more idiomatic:
`raise UnsupportedOperatorError(...)`

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks for this change!

While overall I think this is a big improvement, I'd like to discuss an alternative.
Maintaining a global singleton _symbolic_function_state, and having symbolic functions get access to it through a global getter function seems a lot less clean than passing it in as an argument. I find it's generally more obvious for functions to depend on their inputs rather than global state, and we should aim to make the code as obvious as possible.

What do you think of changing all of the symbolic function signatures to take an extra argument containing the state?

My main concern with this is how do we transition the custom symbolic functions that are not in this repo.

Maybe we could do something like:
```python
try:
   symbolic_fn(symbolic_state, g, *inputs, **attrs)
except TypeError as e:
   if is_too_many_positional_args_error(e):
     logging.warn("<function name> is using the old signature. It should be updated to take a symbolic function state as its first argument. This will become a fatal error in the next release of PyTorch")
  symbolic_fn(g, *inputs, **attrs)
```
We can discuss in person if you prefer.

Unrelated:

> > As a curiosity what is the motivation for such change?
> 
> Better maintainability ...

@BowenBao can you please add this rationale to the PR description?
In general PR descriptions should contain the rationale. It's often very helpful for current and future readers.

File: docs/source/onnx.rst [hunk 13]
```diff
@@ -342,9 +342,15 @@ If the operator is an ATen operator (shows up in the TorchScript graph with the
   Make sure the function has the same name as the ATen function, which may be declared in
   ``torch/_C/_VariableFunctions.pyi`` or ``torch/nn/functional.pyi`` (these files are generated at
   build time, so will not appear in your checkout until you build PyTorch).
-* The first arg is always the ONNX graph that is being built for export.
+* The first arg is always the ONNX graph that is being built for export (except for advanced symbolic functions).
```
Comment (garymm) [inline docs/source/onnx.rst:345]: I'd reword to
"By default, the first arg is the ONNX graph...". Don't mention the advanced case in this bullet point.

File: docs/source/onnx.rst [hunk 14]
```diff
@@ -342,9 +342,15 @@ If the operator is an ATen operator (shows up in the TorchScript graph with the
...
   Other arg names must EXACTLY match the names in the ``.pyi`` file,
   because dispatch is done with keyword arguments.
+* Advanced symbolic functions may accept an additional `torch.onnx.SymbolicContext`
```
Comment (garymm) [inline docs/source/onnx.rst:348]: Suggested wording:

A symbolic function that has a first arg (before the ``Graph`` object) with the type annotation of ``torch.onnx.SymbolicContext`` will be called with that additional context. See examples below.

File: torch/onnx/__init__.py [hunk 15]
```diff
@@ -23,6 +24,15 @@ class CheckerError(Exception):
     pass
 
 
+class SymbolicContext:
+    def __init__(self, params_dict, env, cur_node, onnx_block):
+        self.params_dict: Optional[Dict[str, _C.IValue]] = params_dict
```
Comment (garymm) [inline torch/onnx/__init__.py]: Do these need to be Optional now that we're not using a single global object?

File: torch/onnx/symbolic_helper.py [hunk 16]
```diff
@@ -46,16 +45,21 @@
 # more robust to different graphs.
 #
 # ----------------------------------------------------------------------------------
-# Extra context for Symbolic functions
+# Advanced symbolic functions
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:48]: I like the "Extra context for symbolic functions" title better.

File: torch/onnx/utils.py [hunk 17]
```diff
@@ -1015,6 +1015,15 @@ def _should_aten_fallback(ns, op_name, opset_version, operator_export_type):
...
 
 
+def _need_symbolic_context(symbolic_fn):
+    # Check if the first argument to symbolic_fn is annotated as type `torch.onnx.SymbolicContext`
+    full_arg_spec = inspect.getfullargspec(symbolic_fn)
+    args = full_arg_spec.args
+    annotations = full_arg_spec.annotations
+    if len(args) == 0:
+        return False
+    return args[0] in annotations and annotations[args[0]] == SymbolicContext
```
Comment (garymm) [inline torch/onnx/utils.py:1025]: Probably not very important, but I suggest allowing subclasses:
`and issubclass(annotations[args[0]], SymbolicContext)`

File: torch/onnx/utils.py [hunk 18]
```diff
@@ -1015,6 +1015,15 @@ def _should_aten_fallback(ns, op_name, opset_version, operator_export_type):
     return is_onnx_aten_export or (not is_exportable_aten_op and is_aten_fallback_export)
 
 
+def _need_symbolic_context(symbolic_fn):
+    # Check if the first argument to symbolic_fn is annotated as type `torch.onnx.SymbolicContext`
+    full_arg_spec = inspect.getfullargspec(symbolic_fn)
```
Comment (garymm) [inline torch/onnx/utils.py:1020]: Up to you, but maybe slightly more compact and just as readable way of getting the same thing:

```python
params = inspect.signature(symbolic_fn).parameters.values()
return params and issubclass(params[0].annotation, SymbolicContext)
```

It's nice that this API doesn't require us to do the mapping between args and annotations.

File: torch/onnx/symbolic_helper.py [hunk 19]
```diff
@@ -42,6 +43,23 @@
...
+#
+# ----------------------------------------------------------------------------------
+# Advanced symbolic functions
+# ----------------------------------------------------------------------------------
+#
+# In general, symbolic functions only require inputs and attributes to
+# the original node. In rare circumstances, extra context may be required.
+# For example, symbolic function for `prim::Loop` needs access to the subblock of
+# the original node.
+# For these advanced symbolic functions, an additional `torch.onnx.SymbolicContext`
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:55]: Same suggested wording edit from the onnx.rst file applies here.

**Review by garymm — APPROVED** [review]
Thanks!

File: torch/onnx/__init__.py [hunk 20]
```diff
@@ -25,8 +25,12 @@ class CheckerError(Exception):
 
 
 class SymbolicContext:
-    r"""
-    Provides extra context for advanced symbolic functions.
+    r"""Provides extra context for symbolic functions.
+    Args:
+        params_dict (Dict[str, _C.IValue]): Mapping from graph initializer name to IValue.
```
Comment (garymm) [inline torch/onnx/__init__.py:30]: Please put the type annotations into the signature of __init__ rather than the doc string. That way they're accessible to `inspect` and mypy.

Comment (BowenBao) [reply torch/onnx/__init__.py:30]: I'm trying to see how to pass the new CI test on doc that was recently added. Saw another example doing it this way, seems like the sphinx standard. But no luck passing it yet.
